### ISSUE TITLE
Nx does not ignore package.json based on workspace settings

### ISSUE THREAD

**Issue by JacobLey** [nrwl/nx#22348]
### Current Behavior

`pnpm-workspace.yaml`:
```
packages:
  - apps/**
  - '!apps/example/src/tests/**'
```

A `package.json` file contained at `apps/example/src/tests/data/package.json` is included as a project, despite pnpm explicitly ignoring it.

### Expected Behavior

The package.json (which only contains `{ "type": "commonjs" }`) should be ignored by Nx

### Steps to Reproduce

1. Create a package.json file anywhere in your repo that does not already have it
2. Ignore the file in `pnpm-workspace.yaml` 
3. Run any pnpm script, and inspect the `context.projectsConfigurations.projects` property, which will contain the package.json transformed to a project.

### Nx Report

```shell
Node   : 20.11.1
OS     : linux-arm64
pnpm   : 8.15.4

nx (global)  : 18.1.1
nx           : 18.1.1
@nx/eslint   : 18.1.1
@nrwl/tao    : 18.1.1
```

### Package Manager Version

pnpm 8.15.4

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

Issue appears to have been introduced in this PR: https://github.com/nrwl/nx/pull/19983

Unless I am mistaken, the offending logic is here: https://github.com/nrwl/nx/blob/master/packages/nx/src/plugins/package-json-workspaces/create-nodes.ts#L42

The _intention_ of the code appears to separate the positive and negative matches.
First find all packages that explicitly match the package.
Then use the negative matcher to exclude packages.

However it doesn't appear the negative matcher works as intended. The packages to ignore are still picked up, and are then a "node" is created from the package.json.

That `!negativePatterns.some((negative) => minimatch(p, negative))` also seems like incorrect logic to me.
If I have two (or more) negative patterns that ignore separate locations in my codebase, a `package.json` that is ignored by only one of them (and therefore should be totally ignored) would still end up getting transformed.

Although as I have tested, regardless of the response of that extra step, it is not enough to exclude the packages altogether.

The justification for this logic seems to be coming from here: https://github.com/nrwl/nx/issues/19225

Which suggests to ignore the negative matches altogether. It seems the problem arises from computing the glob patterns individually (which would cause a negative to match nearly everything). Instead they should be computed only against the successful matches from the positives.

**Comment by JacobLey** [comment]
An important callout is that a `pnpm-workspace.yaml` that looks like:
```
packages: 
  - foo/**
  - !bar/**
```

Would work as expected with existing logic. That is because`foo/` is mutually exclusive from `bar/`. The positive matchers for foo do not pick up bar.

However in the problem case, the excluded `package.json`s are a subset of the main directory. The the positive matchers pick up the _should be ignored_ `package.json`

**Comment by matheo** [comment]
Thanks for reporting this, I'm suffering the same issue :)

**Comment by matheo** [comment]
@AgentEnder Adding another related issue to `pnpm`
as I'm using v9 and this issue is back: https://github.com/nrwl/nx/pull/17407
I cannot build a NestJS application generating the package.json file

I see that this repo is fixed to `pnpm v6` lockfile and not using the latest v9 :thinking:

**Comment by matheo** [comment]
I just realized that `v9` lockfile is not supported yet and I might recommend to put `engines` `pnpm < 9` in the `nx` package.json so we don't try it before Nx parses the lock-file properly.

**Comment by JacobLey** [comment]
It appears the code I called out above was fixed in https://github.com/nrwl/nx/commit/28b3d80f2f1560fb3dfe720e919087964c6a85b3

However I am still experiencing this issue.

From what I can tell, that `createNodes` is not the only source-of-truth for finding packages.

It appears that this plugin still picks up rogue files, and doesn't fully enforce negated configs: https://github.com/nrwl/nx/blob/65c3e560e28bf6683d1b1d0e119993ad70e015ef/packages/nx/src/plugins/target-defaults/target-defaults-plugin.ts#L39

`minimatch('a','{a,!a}')` would be true, despite include a negation for `a`.

However that was removed in a recent commit: https://github.com/nrwl/nx/commit/2ae57dcfbea17acfc97e8fec21c71e9ff2d2b295#diff-f18a077abebe049eeacc637f1723c2b6214849af9e112c86042ebf8fa4e9f1c9 just last week

So I'm hoping the next release (~9.5) will magically resolve this issue?
